File: problems/3-Longest-Substring-Without-Repeating-Characters/cpp/Solution.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        if (s.length() == 0) {
            return 0;
        }
        int map[256];
        int p = 0;
        int q = 1;
        int max = 0;
        memset(map, -1, sizeof(map));
        map[s[0]] = 0;
        while (q < s.length()) {
            if (map[s[q]] == -1) {
                map[s[q]] = q;
                q ++;
            } else {
                if (max < q - p) {
                    max = q - p;
                }
                p = map[s[q]] + 1;
                memset(map, -1, sizeof(map));
                map[s[p]] = p;
                q = p + 1;
            }
        }
        if (max < q - p) {
            max = q - p;
        }
        return max;
    }
};

File: problems/3-Longest-Substring-Without-Repeating-Characters/java/Solution.java
public class Solution {
    public int lengthOfLongestSubstring(String s) {
        int len = s.length();
        int max = 0;
        int i = 0;
        Map<Character, Integer> lastIndex = new HashMap<>();
        while (i < len) {
            lastIndex.put(s.charAt(i), i);
            int j = i + 1;
            while (j < len && !lastIndex.containsKey(s.charAt(j))) {
                lastIndex.put(s.charAt(j), j);
                j++;
            }
            if (max < j - i) {
                max = j - i;
            }
            if (j == len) {
                break;
            }
            i = lastIndex.remove(s.charAt(j)) + 1;
            lastIndex.clear();
        }
        return max;
    }
}
